### ISSUE TITLE
iOS/macOS passkeys not backup eligible

### ISSUE THREAD

**Issue by videah** [kanidm/webauthn-rs#185]
After implementing a server using this crate I'm finding that the `backup_eligible` bit is set to false for iOS/macOS passkeys, giving me a `CredentialBackupElligibilityInconsistent` error and making them unusable on anything other than the device used to generate them.

Not sure if something needs to be done on my end or if this is unimplemented functionality/a bug.

**Comment by Firstyear** [comment]
@videah Trying to understand this a bit better. What iOS/macOS versions are you using? What exact steps reproduce? From the report I am guessing something like:

* enroll a passkey on macOS
* authenticate with the passkey on macOS (works)
* authenticate with the passkey on iOS (doesn't work)

Is that correct?

Can you also give extra detail with tracing output (debug or trace level)

Thanks!

**Comment by videah** [comment]
I am using the latest iOS 16 public beta and macOS 12.4 with the developer preview of passkeys enabled in Safari. The steps you listed are correct, it occurs when done vice versa (enrolling on iOS) as well.

Here is my tracing output:

```
2022-08-28T14:25:16.948630Z TRACE webauthn_rs_core::core: state=RegistrationState { policy: Preferred, exclude_credentials: [], challenge: Base64UrlSafeData([105, 237, 99, 207, 8, 143, 157, 45, 208, 37, 0, 61, 164, 253, 95, 184, 251, 103, 41, 199, 167, 108, 148, 169, 8, 9, 156, 36, 204, 250, 125, 145]), credential_algorithms: [ES256, RS256], require_resident_key: false, authenticator_attachment: None, extensions: RequestRegistrationExtensions { cred_protect: None, cred_blob: None, uvm: Some(true), cred_props: Some(true), min_pin_length: None, hmac_create_secret: None }, experimental_allow_passkeys: true }
2022-08-28T14:25:16.948727Z TRACE webauthn_rs_core::core: reg=RegisterPublicKeyCredential { id: "i_9Vn1iXd2u47qJ4n_RaNqzzx3c", raw_id: Base64UrlSafeData([139, 255, 85, 159, 88, 151, 119, 107, 184, 238, 162, 120, 159, 244, 90, 54, 172, 243, 199, 119]), response: AuthenticatorAttestationResponseRaw { attestation_object: Base64UrlSafeData([163, 99, 102, 109, 116, 100, 110, 111, 110, 101, 103, 97, 116, 116, 83, 116, 109, 116, 160, 104, 97, 117, 116, 104, 68, 97, 116, 97, 88, 152, 145, 32, 176, 28, 9, 255, 201, 110, 127, 7, 27, 186, 230, 218, 177, 139, 94, 31, 31, 108, 118, 193, 197, 51, 238, 82, 174, 119, 137, 180, 146, 111, 69, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 20, 139, 255, 85, 159, 88, 151, 119, 107, 184, 238, 162, 120, 159, 244, 90, 54, 172, 243, 199, 119, 165, 1, 2, 3, 38, 32, 1, 33, 88, 32, 43, 87, 214, 88, 155, 157, 136, 196, 173, 222, 198, 45, 32, 241, 90, 228, 128, 167, 175, 118, 43, 144, 84, 231, 170, 7, 127, 227, 135, 58, 209, 0, 34, 88, 32, 137, 223, 210, 153, 215, 59, 11, 51, 236, 22, 25, 214, 208, 33, 92, 8, 13, 88, 60, 126, 51, 70, 153, 240, 218, 8, 139, 218, 105, 115, 121, 101]), client_data_json: Base64UrlSafeData([123, 34, 116, 121, 112, 101, 34, 58, 34, 119, 101, 98, 97, 117, 116, 104, 110, 46, 99, 114, 101, 97, 116, 101, 34, 44, 34, 99, 104, 97, 108, 108, 101, 110, 103, 101, 34, 58, 34, 97, 101, 49, 106, 122, 119, 105, 80, 110, 83, 51, 81, 74, 81, 65, 57, 112, 80, 49, 102, 117, 80, 116, 110, 75, 99, 101, 110, 98, 74, 83, 112, 67, 65, 109, 99, 74, 77, 122, 54, 102, 90, 69, 34, 44, 34, 111, 114, 105, 103, 105, 110, 34, 58, 34, 104, 116, 116, 112, 115, 58, 47, 47, 109, 97, 99, 98, 111, 111, 107, 46, 119, 101, 114, 101, 119, 111, 108, 102, 45, 97, 117, 103, 109, 101, 110, 116, 101, 100, 46, 116, 115, 46, 110, 101, 116, 34, 125]), transports: None }, type_: "public-key", extensions: RegistrationExtensionsClientOutputs { appid: None, cred_blob: None, cred_props: None } }
2022-08-28T14:25:16.949358Z TRACE webauthn_rs_core::internals: AttestationObjectInner v=Map({Text("fmt"): Text("none"), Text("attStmt"): Map({}), Text("authData"): Bytes([145, 32, 176, 28, 9, 255, 201, 110, 127, 7, 27, 186, 230, 218, 177, 139, 94, 31, 31, 108, 118, 193, 197, 51, 238, 82, 174, 119, 137, 180, 146, 111, 69, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 20, 139, 255, 85, 159, 88, 151, 119, 107, 184, 238, 162, 120, 159, 244, 90, 54, 172, 243, 199, 119, 165, 1, 2, 3, 38, 32, 1, 33, 88, 32, 43, 87, 214, 88, 155, 157, 136, 196, 173, 222, 198, 45, 32, 241, 90, 228, 128, 167, 175, 118, 43, 144, 84, 231, 170, 7, 127, 227, 135, 58, 209, 0, 34, 88, 32, 137, 223, 210, 153, 215, 59, 11, 51, 236, 22, 25, 214, 208, 33, 92, 8, 13, 88, 60, 126, 51, 70, 153, 240, 218, 8, 139, 218, 105, 115, 121, 101])})

2022-08-28T14:25:16.949525Z TRACE webauthn_rs_core::internals: extensions=None
2022-08-28T14:25:16.949569Z DEBUG webauthn_rs_core::core: extensions: RegistrationSignedExtensions { cred_protect: None, hmac_secret: None, unknown_keys: {} }
2022-08-28T14:25:16.949594Z DEBUG webauthn_rs_core::core: attestation is: None
2022-08-28T14:25:16.949615Z DEBUG webauthn_rs_core::core: attested credential data is: AttestedCredentialData { aaguid: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], credential_id: Base64UrlSafeData([139, 255, 85, 159, 88, 151, 119, 107, 184, 238, 162, 120, 159, 244, 90, 54, 172, 243, 199, 119]), credential_pk: Map({Integer(1): Integer(2), Integer(3): Integer(-7), Integer(-1): Integer(1), Integer(-2): Bytes([43, 87, 214, 88, 155, 157, 136, 196, 173, 222, 198, 45, 32, 241, 90, 228, 128, 167, 175, 118, 43, 144, 84, 231, 170, 7, 127, 227, 135, 58, 209, 0]), Integer(-3): Bytes([137, 223, 210, 153, 215, 59, 11, 51, 236, 22, 25, 214, 208, 33, 92, 8, 13, 88, 60, 126, 51, 70, 153, 240, 218, 8, 139, 218, 105, 115, 121, 101])}) }
2022-08-28T14:25:16.950568Z TRACE webauthn_rs_core::internals: extensions=RegisteredExtensions { cred_protect: NotRequested, hmac_create_secret: NotRequested, appid: NotRequested, cred_props: Ignored }
2022-08-28T14:25:16.950590Z DEBUG webauthn_rs_core::core: attested_ca_crt = None
```

**Comment by Firstyear** [comment]
I think the issue is that you have macOS 12.4 here with safari tech preview. The "sync" stuff I think is finalised in macOS 13, which is likely why you are seeing an inconsistency between the backup enablement bits. So webauthn-rs is correctly reporting the situation, and correctly raising the security error because a change to backup enablement affects risk assessment of the credential.

So I think this will be resolved in macOS 13, but other than that, I think it's a bug in macOS/iOS itself. Sadly the passkey sync is still a preview, so I was expecting "rough edges" like this :( 

 The options to proceed are:

* Report the issue to Apple - it's unlikely they will fix it in macOS 12 however ...
* Loosen the check in webauthn-rs - I really don't want to do this, because this is a genuine security alert that the credential has altered it's security properties. 
* Wait for a bit until the final versions for macOS/iOS are released out of dev/tech preview.

**Comment by videah** [comment]
I did originally think the same but I'm not sure macOS is at fault because passkeys enrolled on iOS 16 also have `backup_eligible` set to false, so trying to use passkeys that were registered with my iPhone on my iPad produces the same result. I have a friend who has access to macOS 13 so later today I'll try and see if it changes anything but I think there's something else at fault.

**Comment by Firstyear** [comment]
It is a bug in macOS/iOS:

https://github.com/w3c/webauthn/blob/main/index.bs#L983

"Backup eligibility is a [=credential property=] and is permanent for a given [=public key credential source=]."

vs Backup State which shows the current status of the backup/sync of the credential and MAY change. 

So as I'm reading it, this is a fault in macOS/iOS not keeping the BE bit permanently. I'm going to submit this as an issue to the Webauthn WG.

**Comment by Firstyear** [comment]
https://github.com/w3c/webauthn/issues/1791

**Comment by Firstyear** [comment]
WG agrees it's an Apple bug, and that RP's should be able to reject the keys in this case and we need better guidance in the spec. 

The only worry I'd have here is when credentials are "upgraded" from a legacy passkey (without be=true) to a new one. I'm wondering if we make it so in the "accept passkeys" bool flag, we allow the BE state to change in one direction (false to true) but not the other way? Doesn't quite resolve your issue today but it will solve the issues with future enrollments.

**Comment by Firstyear** [comment]
@videah fixed in 4.6, passkeys allow one way upgrades.
